### PR TITLE
[Altair] Lots of test refactors

### PR DESCRIPTION
**Pull request by realbigsean**
Got rid of the `TestingXBuilders` and got most things passing with phase0, haven't tested with altair yet
- Op pool tests
- State transition vector tests (only half passing so far)
- Per block processing tests

### CODE REVIEW

File: beacon_node/beacon_chain/src/lib.rs [hunk 1]
```diff
@@ -1,4 +1,30 @@
...
+
+pub use self::beacon_chain::{
+    AttestationProcessingOutcome, BeaconChain, BeaconChainTypes, BeaconStore, ChainSegmentResult,
+    ForkChoiceError, StateSkipConfig, MAXIMUM_GOSSIP_CLOCK_DISPARITY,
+};
+pub use self::beacon_snapshot::BeaconSnapshot;
+pub use self::chain_config::ChainConfig;
+pub use self::errors::{BeaconChainError, BlockProductionError};
+
+// For lazy-static
```
Comment (michaelsproul) [inline beacon_node/beacon_chain/src/lib.rs]: This is a rogue comment (should be on the `recursion_limit` at the start)

File: Cargo.lock [hunk 2]
```diff
@@ -1699,7 +1699,8 @@ dependencies = [
 [[package]]
 name = "discv5"
 version = "0.1.0-beta.3"
-source = "git+https://github.com/sigp/discv5?rev=02d2c896c66f8dc2b848c3996fedcd98e1dfec69#02d2c896c66f8dc2b848c3996fedcd98e1dfec69"
+source = "registry+https://github.com/rust-lang/crates.io-index"
```
Comment (michaelsproul) [inline Cargo.lock]: Using a nightly compiler generates noise like this in the `Cargo.lock`, would you mind recompiling with stable? (that should revert it)

File: beacon_node/operation_pool/src/lib.rs [hunk 3]
```diff
@@ -500,117 +500,111 @@ impl<T: EthSpec + Default> PartialEq for OperationPool<T> {
...
+                })
+                .unwrap();
 
             assert_eq!(
                 att1.aggregation_bits.num_set_bits(),
-                earliest_attestation_validators(&att1, state).num_set_bits()
+                earliest_attestation_validators(&att1, &state).num_set_bits()
             );
+
+            //TODO: handle altair
```
Comment (michaelsproul) [inline beacon_node/operation_pool/src/lib.rs:605]: I'm happy to update these tests for Altair, seeing as I implemented the other Altair op pool changes

File: consensus/state_processing/src/per_block_processing.rs [hunk 4]
```diff
@@ -22,8 +22,6 @@ pub use verify_deposit::{
 pub use verify_exit::{verify_exit, verify_exit_time_independent_only};
 
 pub mod altair;
-// FIXME(altair): re-enable
-// pub mod block_processing_builder;
```
Comment (michaelsproul) [inline consensus/state_processing/src/per_block_processing.rs:26]: :tada:

File: consensus/state_processing/src/per_block_processing/tests.rs [hunk 5]
```diff
@@ -1,25 +1,72 @@
...
+            Hash256::zero(),
+            (1..last_slot_of_epoch.as_u64())
+                .map(Slot::new)
+                .collect::<Vec<_>>()
+                .as_slice(),
+            (0..num_validators).collect::<Vec<_>>().as_slice(),
+        );
+    }
+    harness
+}
```
Comment (michaelsproul) [inline consensus/state_processing/src/per_block_processing/tests.rs:56]: I was going to say we should let all of these tests go... but seeing as you've already slogged through updating them I think we should keep them! Sorry about that :flushed: (they mostly duplicate functionality from the spec tests IMO).

File: consensus/types/src/beacon_state/tests.rs [hunk 6]
```diff
@@ -1,24 +1,46 @@
...
+use beacon_chain::test_utils::{BeaconChainHarness, EphemeralHarnessType};
+use beacon_chain::types::{
+    eth_spec, BeaconState, BeaconStateError, ChainSpec, CloneConfig, Domain, Epoch,
+    EthSpec as EthSpec2, FixedVector, FoundationBeaconState, Hash256, Keypair, RelativeEpoch, Slot,
+};
 use std::ops::Mul;
+use swap_or_not_shuffle::compute_shuffled_index;
 
-ssz_and_tree_hash_tests!(FoundationBeaconState);
+// ssz_and_tree_hash_tests!(FoundationBeaconState);
```
Comment (michaelsproul) [inline consensus/types/src/beacon_state/tests.rs]: I think we can safely delete this. The spec tests include SSZ tests now and we also have the fuzzers

File: consensus/types/src/beacon_state/tests.rs [hunk 7]
```diff
@@ -68,31 +90,23 @@ fn test_beacon_proposer_index<T: EthSpec>() {
 
 #[test]
 fn beacon_proposer_index() {
-    test_beacon_proposer_index::<MinimalEthSpec>();
+    test_beacon_proposer_index::<eth_spec::MinimalEthSpec>();
```
Comment (michaelsproul) [inline consensus/types/src/beacon_state/tests.rs]:
```suggestion
    test_beacon_proposer_index::<MinimalEthSpec>();
```

Can drop the `eth_spec::` throughout

**Review by michaelsproul — COMMENTED** [review]
Great work! You've covered a lot of ground in this PR

Sorry for jumping in with a review while this is still marked WIP, feel free to disregard some of my more nitpicky comments :sweat_smile: 

I'm keen to get this merged to `altair` soon, and I'm happy to jump in to help finish it off if you throw a few TODOs my way.

With the beacon chain tests, I think we can merge to `altair` without deciding on a fork strategy yet. I've merged the deletion of the global fork schedule, so _fingers crossed_ the tests just run with the `init_fork_schedule` stuff deleted. Later we can bring in an environment variable and something like the `testing_spec` that I've used in the EF tests.

I'm a bit stumped by the 256-epoch advance that we need to do for the exit tests. I'll think about that some more (starting from epoch 256 would be ideal but I'm not sure if that's feasible with the harness).

File: consensus/state_processing/src/per_block_processing/tests.rs (hunk 5 above)
Comment (realbigsean) [reply consensus/state_processing/src/per_block_processing/tests.rs:56]: > they mostly duplicate functionality from the spec tests IMO 

I did notice this once I was about halfway done :joy: 

No problem, if they end up being slow or annoying to maintain, I'm not attached, so would be happy to delete them.

Comment (realbigsean) [reply consensus/state_processing/src/per_block_processing/tests.rs:58]: What are your thoughts on whether the keep the `state_processing/tests/tests.rs` ?

**Comment by realbigsean** [comment]
I appreciate the early review! Fixed the rest of the `state_transition_vector` tests and the tests are all still running.

> I'm a bit stumped by the 256-epoch advance that we need to do for the exit tests. I'll think about that some more (starting from epoch 256 would be ideal but I'm not sure if that's feasible with the harness).

I updated this to start from a slot before of the 256'th epoch and advance a slot. This seems to be necessary for at least setting activation epochs for validators. 

What are your thoughts on whether to keep the tests in `state_processing/tests/tests.rs`?

**Review by michaelsproul — APPROVED** [review]

**Comment by michaelsproul** [comment]
Awesome stuff! I added some simplifications on top in this commit: https://github.com/sigp/lighthouse/pull/2279/commits/610e92b8d5068d658892d3752ca61144128f6cfc

**Pull request merged**